core/shared/src/main/scala/zio/clock/package.scala

```diff
@@ -27,18 +27,41 @@ package object clock {
 
   object Clock extends PlatformSpecific with Serializable {
     trait Service extends Serializable {
+
       def currentTime(unit: TimeUnit): UIO[Long]
+
+      // Could be UIO. We keep IO to preserve binary compatibility.
       def currentDateTime: IO[DateTimeException, OffsetDateTime]
-      def instant: UIO[java.time.Instant]                               = currentTime(TimeUnit.MILLISECONDS).map(java.time.Instant.ofEpochMilli(_))
+
+      // The implementation is only here to preserve binary compatibility.
+      def instant: UIO[java.time.Instant] = currentTime(TimeUnit.MILLISECONDS).map(java.time.Instant.ofEpochMilli(_))
+
+      // This could be a UIO. We keep IO to preserve binary compatibility.
+      // The implementation is only here to preserve binary compatibility.
       def localDateTime: IO[DateTimeException, java.time.LocalDateTime] = currentDateTime.map(_.toLocalDateTime())
+
       def nanoTime: UIO[Long]
+
       def sleep(duration: Duration): UIO[Unit]
     }
 
     object Service {
       val live: Service = new Service {
         def currentTime(unit: TimeUnit): UIO[Long] =
-          IO.effectTotal(System.currentTimeMillis).map(l => unit.convert(l, TimeUnit.MILLISECONDS))
+          instant.map { inst =>
+            // A nicer solution without loss of precision or range would be
+            // unit.toChronoUnit.between(Instant.EPOCH, inst)
+            // However, ChronoUnit is not available on all platforms
+            unit match {
+              case TimeUnit.NANOSECONDS =>
+                val micros = inst.toEpochMilli() * 1000000 + inst.getNano()
+                unit.convert(micros, TimeUnit.NANOSECONDS)
+              case TimeUnit.NANOSECONDS | TimeUnit.MICROSECONDS =>
```

Comment:
Not sure I understand this line. If we matched on `NANOSECONDS` above shouldn't we just be matching on `MICROSECONDS` here?

Reply:
Good catch. That's a leftover. Initially I only had the second case, but the decided to add the first one in order to keep the loss in range and precision as low as possible.